misc/testerina/modules/testerina-core/src/main/ballerina/annotation_processor.bal

```diff
@@ -46,26 +46,55 @@ public function registerTest(string name, function f) {
+        boolean isTestFunctionIsolated = f is isolated function;
+        boolean isDataProviderIsolated = true;
+        boolean isTestFunctionParamSafe = true;
+        boolean isSatisfiedParallelizableConditions = isTestFunctionIsolated;
+        string[] reasonToSerialExecution = [];
+
+        DataProviderReturnType? & readonly params = ();
         error? diagnostics = ();
         if config.dataProvider != () {
             var providerFn = config.dataProvider;
 
             if providerFn is function () returns (DataProviderReturnType?) {
+                isDataProviderIsolated = (<function>providerFn is isolated function);
+                isTestFunctionParamSafe = isFunctionParamConcurrencySafe(f);
+                isSatisfiedParallelizableConditions = isTestFunctionIsolated && isDataProviderIsolated && isTestFunctionParamSafe;
                 DataProviderReturnType providerOutput = providerFn();
-                params = providerOutput;
+                params = <DataProviderReturnType & readonly>providerOutput;
             } else {
                 diagnostics = error("Failed to execute the data provider");
             }
         }
+
+        // Register the reason for serial execution.
+        if !isTestFunctionIsolated {
+            reasonToSerialExecution.push("non-isolated test function");
+        }
+
+        if !isDataProviderIsolated {
+            reasonToSerialExecution.push("non-isolated data-provider function");
+        }
+
+        if !isTestFunctionParamSafe {
+            reasonToSerialExecution.push("unsafe test parameters");
+        }
+
+        // If the test function is not parallelizable, then print the reason for serial execution.
+        if !isSatisfiedParallelizableConditions && !config.serialExecution && (conMgr.getConfiguredWorkers() > 1) {
+            println("WARNING: Test function '" + name + "' cannot be parallelized due to " + string:'join(",", ...reasonToSerialExecution));
+        }
+
         boolean enabled = config.enable && (filterGroups.length() == 0 ? true : hasGroup(config.groups, filterGroups))
             && (filterDisableGroups.length() == 0 ? true : !hasGroup(config.groups, filterDisableGroups)) && hasTest(name);
         config.groups.forEach('group => groupStatusRegistry.incrementTotalTest('group, enabled));
 
         testRegistry.addFunction(name = name, executableFunction = f, params = params, before = config.before,
-            after = config.after, groups = config.groups, diagnostics = diagnostics, dependsOn = config.dependsOn,
-            enabled = enabled, dependsOnCount = config.dependsOn.length(), config = config);
-        return true;
+            after = config.after, groups = config.groups.cloneReadOnly(), diagnostics = diagnostics, dependsOn = config.dependsOn.cloneReadOnly(),
+            parallelizable = (!config.serialExecution && isSatisfiedParallelizableConditions && (conMgr.getConfiguredWorkers() > 1)), config = config.cloneReadOnly());
```

Comment:
Wouldn't it be better to create a variable for `(!config.serialExecution && isSatisfiedParallelizableConditions && (conMgr.getConfiguredWorkers() > 1))` and pass it here for the clarity of the code

Reply:
I also thought that initially. But, when returning from a function, we just return it without creating a variable. Isn't it a similar situation?